framework/src/mesh/SpiralAnnularMesh.C

```diff
@@ -0,0 +1,306 @@
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "num_rings", "num_rings>0", "The number of rings.");
+  params.addParam<boundary_id_type>(
+      "cylinder_bid", 1, "The boundary id to use for the cylinder (inner circle)");
+  params.addParam<boundary_id_type>(
+      "exterior_bid", 2, "The boundary id to use for the exterior (outer circle)");
+  params.addParam<Real>("initial_delta_r",
+                        "Width of the initial layer of elements around the cylinder."
+                        "This number should be approximately"
+                        " 2 * pi * inner_radius / nodes_per_ring to ensure that the"
+                        " initial layer of elements is almost equilateral");
+  params.addClassDescription("Creates an annual mesh based on TRI3 elements"
+                             " (it can also be TRI6 elements) on several rings.");
+
+  return params;
+}
+
+SpiralAnnularMesh::SpiralAnnularMesh(const InputParameters & parameters)
+  : MooseMesh(parameters),
+    _inner_radius(getParam<Real>("inner_radius")),
+    _outer_radius(getParam<Real>("outer_radius")),
+    _radial_bias(1.0),
+    _nodes_per_ring(getParam<unsigned int>("nodes_per_ring")),
+    _second_order(getParam<bool>("my_second_order")),
+    _num_rings(getParam<unsigned int>("num_rings")),
+    _cylinder_bid(getParam<boundary_id_type>("cylinder_bid")),
+    _exterior_bid(getParam<boundary_id_type>("exterior_bid")),
+    _initial_delta_r(2 * libMesh::pi * _inner_radius / _nodes_per_ring)
+{
+
+  // catch likely user errors
+  if (_outer_radius <= _inner_radius)
+    mooseError("SpiralAnnularMesh: outer_radius must be greater than inner_radius");
+  if (_num_rings == 0)
+    mooseError("Cannot create mesh with _num_rings == 0");
+}
+
+std::unique_ptr<MooseMesh>
+SpiralAnnularMesh::safeClone() const
+{
+  return libmesh_make_unique<SpiralAnnularMesh>(*this);
+}
+
+void
+SpiralAnnularMesh::buildMesh()
+{
+  {
+
+    // Compute the radial bias given:
+    // .) the inner radius
+    // .) the outer radius
+    // .) the initial_delta_r
+    // .) the desired number of intervals
+    // Note: the exponent n used in the formula is one less than the
+    // number of rings the user requests.
+    Real alpha = 1.1;
+    int n = _num_rings - 1;
+
+    // lambda used to compute the residual and Jacobian for the Newton iterations.
+    // We capture parameters which don't need to change from the current scope at
+    // the time this lambda is declared. The values are not updated later, so we
+    // can't use this for e.g. f, df, and alpha.
+    auto newton = [this, n](Real & f, Real & df, const Real & alpha) {
+      f = (1. - std::pow(alpha, n + 1)) / (1. - alpha) -
+          (_outer_radius - _inner_radius) / _initial_delta_r;
+      df = (-(n + 1) * (1 - alpha) * std::pow(alpha, n) + (1. - std::pow(alpha, n + 1))) /
+           (1. - alpha) / (1. - alpha);
+    };
+
+    Real f, df;
+    newton(f, df, alpha);
+
+    while (std::abs(f) > 1.e-9)
```

Comment:
It's possible that this Newton iteration could fail to converge (thus making this an infinite while loop). This has not happened yet in my testing but we should guard against it by specifying a maximum number of allowed iterations (say 25) and if that is exceeded throw a `mooseError`.